Newsletters: add 144 (2021-04-14)

harding opened the pull request:
- [x] Add news @harding 
- [x] Update releases/RCs @harding 
- [x] Add topic links @harding 
- [x] Bitcoin Core PR Review Club summary @jonatack
- [x] `Bitcoin Core #21594` rpc: add network field to getnodeaddresses @adamjonas
- [x] `Bitcoin Core #21166` Introduce DeferredSignatureChecker and have SignatureExtractorClass subclass it @jnewbery
- [x] `LND #5108` channeldb+invoices: add spontaneous AMP receiving + sending via SendToRoute @dongcarl

jnewbery commented on _posts/en/newsletters/2021-04-14-newsletter.md:
```diff
@@ -0,0 +1,213 @@
+  of [BIP8][].  The main technical difference between these PRs is how
+  their start and stop points were specified.  PR#21377 uses Median Time
+  Past (MTP)---the median time declared in the headers of the current
+  and previous 11 blocks.  PR#21392 uses the *height* of the current
```
I don't think height (or genesis block) need to be italicized. I expect readers to be familiar with those terms.

jnewbery replied:
```suggestion
  and previous 10 blocks.  PR#21392 uses the *height* of the current
```

The median time past is the median time of "the last 11 blocks", or "this block and its 10 predecessor blocks".

jnewbery commented on _posts/en/newsletters/2021-04-14-newsletter.md:
```diff
@@ -0,0 +1,213 @@
+  their start and stop points were specified.  PR#21377 uses Median Time
+  Past (MTP)---the median time declared in the headers of the current
+  and previous 11 blocks.  PR#21392 uses the *height* of the current
+  block---the number of blocks between it and the *genesis block*, block
```
"between it and the genesis block" doesn't seem right to me. That'd suggest the first block after the genesis block is height zero, since there are no blocks between it and the genesis block.

harding replied:
I agree, this was incorrect.  Removing this clause entirely based on https://github.com/bitcoinops/bitcoinops.github.io/pull/550#discussion_r611781386 (if we expect readers to be familiar with height, we don't need to define it).

jnewbery commented on _posts/en/newsletters/2021-04-14-newsletter.md:
```diff
@@ -0,0 +1,213 @@
+      one set of 2,016 blocks and another set of 2,016 blocks.  It also
+      requires a reorg involving a large number of blocks or for there
+      to have been a recent major change in the amount of PoW required
+      (indicating either a recent major increase or loss of hash rate,
```
s/increase or loss/increase or decrease/ ?

jnewbery commented on _posts/en/newsletters/2021-04-14-newsletter.md:
```diff
@@ -0,0 +1,213 @@
+       This is a change from BIP9 and BIP148 where MTP was used for the
+       *latest* time when counting stopped.  The difference removes the
+       ability for an activation attempt to go directly from *started*
+       to *failed*, simplifying analysis and guaranteeing that miners
+       will get at least one retarget period to activate taproot.
```
Perhaps "miners will get at least one retarget period to activate taproot" -> "there will be at least one retarget period when miners can signal for activation" ?

jnewbery commented on _posts/en/newsletters/2021-04-14-newsletter.md:
```diff
@@ -0,0 +1,213 @@
+    with the compromise, the compromise PR has now received reviews or
+    expressions of support from over a dozen active contributors to
+    Bitcoin Core and the maintainers of two other full node
+    implementations (btcd and libbitcoin).  We hope this momentum
+    to activate taproot continues and we'll be able to report additional
+    progress in a future newsletter.
```
👏

jnewbery commented on _posts/en/newsletters/2021-04-14-newsletter.md:
```diff
@@ -0,0 +1,213 @@
+       a minimum activation height of `0` to activate within the window
+       defined by MTP.
+
+    Although other discussion participants expressed their displeasure
+    with the compromise, the compromise PR has now received reviews or
```
repetition of compromise. Perhaps:

```suggestion
    Although some discussion participants expressed their displeasure
    with the compromise proposal, the implementation has now received reviews or
```

jnewbery reviewed (commented):
Just reviewed the "Taproot activation discussion" and LND 5047 write-up so far.

jnewbery commented on _posts/en/newsletters/2021-04-14-newsletter.md:
```diff
@@ -0,0 +1,225 @@
+[Introduce deploymentstatus][review club
+#19438] is a PR ([#19438][Bitcoin Core #19438]) by Anthony Towns that proposes
+three helper functions to make it easier to [bury future deployments without
+changing all the code paths][easier burying] that check [BIP9][] status:
```
```suggestion
changing all the code paths][easier burying] that check a soft fork's activation status:
```

(the point is to make these functions generic so they don't just apply to bip9 deployments - they also apply to buried deployments)

jnewbery commented on _posts/en/newsletters/2021-04-14-newsletter.md:
```diff
@@ -0,0 +1,225 @@
+     are also buried."
+  a1link="https://bitcoincore.reviews/19438#l-75"
+
+  q2="How many version bits deployments are defined by this PR?"
```
I'm not sure I'd say that these two version bits deployments are "defined by this PR". They are already defined in the code in master.

jonatack replied:
Thanks @jnewbery! You are right, https://github.com/bitcoin/bitcoin/pull/19438/commits/994b0ce58f832da28adf90ab08ce2a7d6886cdae adds the `BuriedDeployment` enum but `DeploymentPos` already exists. That was an oversight on my part.

The question could be instead "How many version bits deployments are there? Where are they listed in the code?" and in the answer, "listed by the PR" replaced by "enumerated in the codebase"

jnewbery replied:
Change looks good to me. Thanks!

jnewbery reviewed (commented):
I've reviewed the parts by @adamjonas and @jonatack. Just a couple of minor comments on each. Great stuff everyone!

jonatack commented on _posts/en/newsletters/2021-04-14-newsletter.md:
```diff
@@ -0,0 +1,225 @@
+  Past ([MTP][]); PR#21392 uses the height of the current block.
+
+    MTP is normally roughly consistent between Bitcoin's main network
+    (mainnet) and its various testnet works, such as testnet, the default
```
"testnet works" almost works, but I gather it was intended to be "test networks"

dongcarl commented:
Talked to Conner and came up with this snippet, let me know if I'm missing any crucial pieces here otherwise perhaps we can just do spontaneous AMP in this week's newsletter!

> LND #5108 adds support for making spontaneous multipath payments (à la Original AMP) using the low-level `sendtoroute` RPC. Original AMP multipath payments are non-interactive (or spontaneous) by nature as the spender selects all preimages. Spender preimage selection is also a part of keysend-style spontaenous payments, which has been used for singlepath spontaenous payments in LND. Followup PRs are expected to make spontaenous multipath payments available to the higher-level `sendpayment` RPC.

jonatack commented on _posts/en/newsletters/2021-04-14-newsletter.md:
```diff
@@ -0,0 +1,225 @@
+     are listed in the `BuriedDeployment` enumerator proposed by the PR in
+     [src/consensus/params.h#L14-22](https://github.com/bitcoin/bitcoin/blob/e72e062e/src/consensus/params.h#L14-L22).
+     One could argue that the [Satoshi-era
+     softforks](/en/topics/soft-fork-activation/#2009-hardcoded-height-consensus-nlocktime-enforcement)
```
correcting myself per the style guide
```suggestion
     soft forks](/en/topics/soft-fork-activation/#2009-hardcoded-height-consensus-nlocktime-enforcement)
```

jonatack commented on _posts/en/newsletters/2021-04-14-newsletter.md:
```diff
@@ -0,0 +1,225 @@
+     activation method, what changes would need to be made to Bitcoin Core, if
+     this PR is merged?"
+  a3="The main change would be greatly simplified compared to the current code:
+     move the `DEPLOYMENT_TAPROOT` line from the `DeploymentPos` enum to the
```
maybe s/enum/enumerator/

jonatack commented on _posts/en/newsletters/2021-04-14-newsletter.md:
```diff
@@ -0,0 +1,225 @@
+BOLTs][bolts repo].*
+
+- [Bitcoin Core #21594][] adds a network field to RPC `getnodeaddresses`
+  to help identify nodes on various networks (i.e. ipv4, ipv6, I2P, onion).
```
maybe best if all the networks are lowercased or normalcased, not sure which
```suggestion
  to help identify nodes on various networks (i.e. IPv4, IPv6, onion, I2P)
```

jonatack commented on _posts/en/newsletters/2021-04-14-newsletter.md:
```diff
@@ -0,0 +1,225 @@
+  to help identify nodes on various networks (i.e. ipv4, ipv6, I2P, onion).
+  The author has also proposed that this sets the groundwork for a future
+  patch for `getnodeaddresses` takes an argument of a specific network
+  and returns only addresses in that network.
```
```suggestion
  and return only addresses in that network.
```

jonatack commented on _posts/en/newsletters/2021-04-14-newsletter.md:
```diff
@@ -0,0 +1,225 @@
+
+- [Bitcoin Core #21594][] adds a network field to RPC `getnodeaddresses`
+  to help identify nodes on various networks (i.e. ipv4, ipv6, I2P, onion).
+  The author has also proposed that this sets the groundwork for a future
```
maybe s/sets/set/ or s/sets/lay/ (laying the groundwork) or could just "that a follow-up patch enable..."

jnewbery replied:
I also hadn't heard "set the groundwork" before, but a google search revealed that it's about as popular as "lay the groundwork", so I didn't comment. I agree that "lay the groundwork" is slightly better.

jonatack replied:
> I also hadn't heard "set the groundwork" before, but a google search revealed that it's about as popular as "lay the groundwork"

Thanks! TIL

jonatack reviewed (commented):
Very timely news! Great job on the taproot activation section.

adamjonas commented on _posts/en/newsletters/2021-04-14-newsletter.md:
```diff
@@ -0,0 +1,225 @@
+    Unfortunately, MTP can be easily manipulated in small ways by a
+    small number of miners and in large ways by a majority of hash rate.
+    It can also revert to an earlier time even by accident during a
+    block chain reorganization (reorg).  By comparison, heights can only
```
Is the (reorg) bit necessary? Seems like that is a common enough reference.

Xekyo reviewed (commented):
Great newsletter. Strong description of the Taproot activation tradeoffs!

harding commented on _posts/en/newsletters/2021-04-14-newsletter.md:
```diff
@@ -0,0 +1,225 @@
+    It can also revert to an earlier time even by accident during a
+    block chain reorganization (reorg).  By comparison, heights can only
+    decrease in extraordinary reorgs,[^height-decreasing] likely
+    requiring the obvious and prolonged manipulation by a majority of
```
Thanks for questioning me on this!  I think you were doubting "requiring [...] manipulation".  In process of writing a defense of that, I realized that there was a way to reduce height that was not "obvious and prolonged", so the concluding clause was flawed anyway.  I'll remove it.

harding commented on _posts/en/newsletters/2021-04-14-newsletter.md:
```diff
@@ -0,0 +1,225 @@
+    These tradeoffs between the two proposals, among other concerns,
+    created an impasse that some developers thought was preventing either
+    PR from receiving additional review and, ultimately, getting one of
+    them merged into Bitcoin Core.  That impasse was resolved to the
```
I think "one of them" is an important part of that clause.

jnewbery replied:
I agree that this should stay.

harding commented:
I believe I've addressed all feedback either through edits or with a reply; thanks everyone for the feedback!  I also reviewed everyone else's contributions; they're great!  I think we could get both MTP and height advocates to agree on one thing: this is one gross newsletter.  :-)

jnewbery commented on _posts/en/newsletters/2021-04-14-newsletter.md:
```diff
@@ -0,0 +1,225 @@
+       ability for an activation attempt to go directly from *started*
+       to *failed*, simplifying analysis and guaranteeing that there
+       will be at least one complete 2,016 block period where miners can
+       signal for activation.
```
New wording seems good to me.

Xekyo replied:
Ah great, now it's much clearer. Thanks!

jnewbery commented on _posts/en/newsletters/2021-04-14-newsletter.md:
```diff
@@ -0,0 +1,227 @@
+  fee][Bitcoin Core #21151].
+
+- [LND #5108][] adds support for making spontaneous [multipath payments][topic
+  multipath payments] (also called Original AMP) using the low-level
```
Perhaps italicize the new terminology:

```suggestion
  multipath payments] (also called _Original AMP_) using the low-level
```

jnewbery commented on _posts/en/newsletters/2021-04-14-newsletter.md:
```diff
@@ -0,0 +1,227 @@
+
+- [LND #5108][] adds support for making spontaneous [multipath payments][topic
+  multipath payments] (also called Original AMP) using the low-level
+  `sendtoroute` RPC. Original AMP are non-interactive (or spontaneous) by nature
```
Is AMP the plural of AMP? I would expect:

```suggestion
  `sendtoroute` RPC. Original AMPs are non-interactive (or spontaneous) by nature
```

jnewbery reviewed (commented):
Looks good to me. All previous review comments have been addressed (thanks @harding!) All new topic links also look good to me.

I have two minor comments on LND 5108.

jonatack commented on _posts/en/newsletters/2021-04-14-newsletter.md:
```diff
@@ -0,0 +1,234 @@
+
+  q1="How many buried deployments are defined by this PR?"
+  a1="Five: height in coinbase, CLTV ([BIP65][] `CHECKLOCKTIMEVERIFY`), strict
+     DER signatures, CSV ([BIP112][] `OP_CHECKSEQUENCEVERIFY`), and segwit. They
```
Yes, good call :+1: also thanks for referring me to BIPs 34 (and 64--is it the right one?)

jonatack replied:
Ah, 66 for the strict DER signatures

harding commented:
@jnewbery thanks for the early morning final review!  Edits pushed.

jonatack commented:
Just checked the edits rendered on localhost and they look good. This feedback by @jnewbery remains, I believe: https://github.com/bitcoinops/bitcoinops.github.io/pull/550#discussion_r612350890. Should I push the proposed change?

jnewbery commented:
> This feedback by @jnewbery remains, I believe: #550 (comment). Should I push the proposed change?

Oops - I thought that'd been addressed. I've changed the text minimally as follows:

```diff
-  q2="How many version bits deployments are defined by this PR?"
+  q2="How many version bits deployments are currently defined?"
```

Look ok to you, @jonatack?

jonatack commented:
@jnewbery yes; also update "listed by the PR"?

jnewbery commented:
@jonatack done!

jonatack commented:
Thanks! 🙏

bitschmidty commented:
💪 newsletter team, thanks to all who participated: @harding @dongcarl @jonatack @jnewbery @adamjonas @Xekyo
